const User = require('../models/User.js');
const filevalidation = require('../service/filevalidation');
const cloudinary = require('cloudinary');
const fs = require('fs-extra');

exports.createUser = async (req, res) => {

	try {
		
		filevalidation.validateFile(req.file); 
		/* 	
			Multer middleware is launched just when a file is sended to our server, 
			because of this we have to validate here that there are some file uploaded
		*/
		
		// Create DB object with the req object data
		const newUser = {};
		const {code, name, contact, lost_at} = JSON.parse(req.body.data);
		
		newUser.code = code;
		newUser.name = name;
		newUser.contact = contact;
		newUser.lost_at = lost_at;
		
		// Upload photo in our cloud
		const result = await cloudinary.v2.uploader.upload(req.file.path);
		
		// Create DB object with the cloudinary object data
		newUser.photo_url = result.secure_url; // Add photo url after the result JSON object is generated by cloudinary
		newUser.public_id = result.public_id; // Add public id after the result JSON object is generated by cloudinary
		const user = new User(newUser);
		
		// Save the data into our DB and delete the saved file
		await user.save();
		await fs.unlink(req.file.path); // Delete the file from our server
		
		// Inform successful data saving
		const SUCCESS_COLOR = "\x1b[1;32m%s\x1b[0m";
		console.log(SUCCESS_COLOR, 'Data saved successfully');
		console.log(SUCCESS_COLOR, result.secure_url);

		res.json({msg: "Data saved successfully"});

	}catch(error){
	/* 
		Error thrown if the file doesn't exist or if something in the database went wrong or if something in the cloud went wrong,
		the data validation and extension validation were made by multer middleware 
	*/
		const ERROR_COLOR = "\x1b[1;31m%s\x1b[0m"; 
		console.log(ERROR_COLOR, "An error has ocurred");
		res.json({error});
	}
}

exports.getUsers = (req, res) => {
	res.send('get');
}
